### PR TITLE
Refresh token flow (and bonus switch accounts)

### LINKED ISSUE #207: Ensure we handle invalid tokens
## Introduction
As of now the application stores the users authentication token after successful login. This is then re-used in future sessions for accessing the API.

Although so far the tokens appear to be long-lived, there will/should be scenarios where the token becomes invalid. At this point we currently have no way to notify the user that the session is invalid and/or offer to log in again.

## Requirements
If the API was to return an unauthorised error when we attempt to use the token we have stored, we should gracefully handle that scenario by attempting to silently login again and/or prompt the user.

### PR DESCRIPTION
**Pull request by mormaer**
# Checklist
- [x] I have read [CONTRIBUTING.md](./CONTRIBUTING.md)
- [x] I have described what this PR contains
- [x] This PR addresses one or more open issues that were assigned to me:
      - fixes #207 
- [x] If this PR alters the UI, I have attached pictures/videos

# Pull Request Information

## About this Pull Request
This PR started off as _"show a modal, get a new token, carry on"_. Unfortunately showing a modal and getting a token was simples, but then came the issue that the entire application was holding onto copies of a now out of date `SavedAccount` with the old token in it.

The token should be separated from the account (I think), but I've not done that in this PR, when we introduce the _middleware_ between the application and the API we can do that bit 🙏 

Onto the changes;

- The accounts page is no longer the root view of the application 🎉 
- The start-up of the application has been adapted and now does one of the following:
    - If the user has a _default account_ (currently this is just the last used) they will go directly to their feed at launch
    - If the user has no _default account_ or no accounts at all, they'll land on the accounts page
- As we now always have an account before entering the main application, the `.currentActiveAccount` in the `AppState` is now a non-optional property 🎉 
- As the `.currentActiveAccount` is now non-optional, we don't need to pass the account into every view, you just ask the `AppState` for it, it'll always be up-to-date and you don't need any unwrapping 🎉 
- All the account passing around has been removed, now everywhere does as above and reads it from the shared `AppState` 🎉 
- If we encounter an invalid session a modal will be presented asking the user to re-authenticate, once done the `.currentActiveAccount` is updated, that in combination with the above should make future interactions succeed with the refreshed token 🎉 
- Added a new `Switch Account` option in settings where the user can switch between any added accounts. When this is selected the whole application UI will reset to give that account a clean slate to start from 🎉 

## Screenshots and Videos

https://github.com/mlemgroup/mlem/assets/5231793/6760eb74-1ac2-445f-aa98-866cfe5ff96a

https://github.com/mlemgroup/mlem/assets/5231793/657d4591-e299-4434-9069-268d6a156100

## Additional Context

A couple of known issues:
- When a user switches account the app state initialises twice very quickly, which causes no _issues_ but we do trigger two `/site` calls as a result, this is more an issue that we should likely be handling that call elsewhere than an issue with switching accounts.
- The modal will not display if we're _already_ inside a modal, this is the age old _you can't present because you're already presenting_ issue, which is easy to get round in `UIKit` but... not so simple in `SwiftUI` unless we want to start manually showing sheets in every modal when we see a token issue 🤮 - I'll continue to look into this and see if we can solve it inside `SwiftUI` 🙏 
- I've kinda borked the iPad navigation (sorry @tht7) we can address that separately since we don't actively support it yet

### CODE REVIEW

**Review by WestonHanners — APPROVED** [review]
This all looks awesome 🚀

File: Mlem/App State.swift [hunk 1]
```diff
@@ -11,20 +11,9 @@ import AlertToast
...
-                            enableDownvote = response.siteView.localSite.enableDownvotes
-                        }
-                    }
-                }
-            }
-        }
-    }
+    @AppStorage("defaultAccountId") var defaultAccountId: Int?
+    @Binding private var selectedAccount: SavedAccount?
+    @Published private(set) var currentActiveAccount: SavedAccount
```
[the remark this answers is not in the document]
Comment (mormaer) [reply Mlem/App State.swift:16]: the `currentActiveAccount` is the non-optional which will always be set if we're inside the main application flow, the `selectedAccount` is a `Binding` to the possibly optional value that lives up at the higher `Window` level of the application.

When a user uses the switch account function we send back the choice using the `selectedAccount` and that causes the main body of the application to be rebuilt for that user, meaning that they have a blank slate as if they've just opened the app for the first time - it saves having to try and reset all the navigation stacks back to that initial state

File: Mlem/Window.swift [hunk 2]
```diff
@@ -8,49 +8,65 @@
...
+                print("Failed while saving data to file: \(writingError)")
             }
-            .onChange(of: favoriteCommunitiesTracker.favoriteCommunities) { newValue in
-                print("Change detected in favorited communities")
+        } catch let encodingError {
+            print("Failed while encoding filters to data: \(encodingError)")
+        }
+    }
+    
+    private func saveFavouriteCommunities(_ newValue: [FavoriteCommunity]) {
```
Comment (EricBAndrews) [inline Mlem/Window.swift:57]: Nit

Favourite -> Favorite

**Review by EricBAndrews — APPROVED** [review]
One small non-blocking nit.

Nice work! This is *so* much cleaner

**Pull request merged**